hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java

```diff
@@ -0,0 +1,190 @@
+		this.connectionProvider = connectionProvider;
+	}
+
+	/**
+	 * Operations can only be added before the {@code SessionFactory} is fully built.
+	 * This is to prevent concurrent additions of operations.
+	 * The operation must use the {@link ServiceRegistry} provided to retrieve
+	 * {@link ConnectionProvider} to get a connection.
+	 *
+	 * The operation will try and be executed eagerly, we might be lucky.
+	 *
+	 * @param operation on a connection ready database
+	 */
+	public void addOperation(Runnable operation) {
+		operations.add( operation );
+		tryAndExecuteOperations();
+	}
+
+	/**
+	 * Try to eagerly flush the operations if a connection is available
+	 */
+	private void tryAndExecuteOperations() {
+		// It would be more elegant to refactor the operation lock from getConnection is a shared method
+		// but I prefer the JVM to have more chances to keep getConnection() inlined
+		try {
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
+			}
+		}
+		catch (Exception e) {
+			// we don't care if there is a failure
+		}
+	}
+
+	@Override
+	public void injectServices(ServiceRegistryImplementor serviceRegistry) {
+		this.serviceRegistry = serviceRegistry;
+		if ( connectionProvider instanceof ServiceRegistryAwareService ) {
+			( (ServiceRegistryAwareService) connectionProvider ).injectServices( serviceRegistry );
+		}
+	}
+
+	@Override
+	public void configure(Map configurationValues) {
+		if ( connectionProvider instanceof Configurable ) {
+			Configurable configurableConnectionProvider = (Configurable) connectionProvider;
+			configurableConnectionProvider.configure( configurationValues );
+		}
+	}
+
+	@Override
+	public Connection getConnection() throws SQLException {
+		Connection connection;
+		try {
+			connection = connectionProvider.getConnection();
+		}
+		catch (Exception e) {
+			throw e;
+		}
+
+		// protect operation execution behind a double checked lock
+		// we don't want them to be applied multiple times
+		if ( operations != null ) {
+			// if lock is already held, we are reentering and acquiring a connection for the an operation execution
+			// so we need to ignore operations here
+			if ( ! lock.isHeldByCurrentThread() ) {
+				lock.lock();
+				try {
+					if ( operations != null ) {
+						// get a service registry bypassing the DelayerConnectionProvider for the action calls
+						//ServiceRegistry sr = buildServiceRegistryDelegator();
+						operations.forEach( Runnable::run );
+						// if an operation fails, the list is not cleared but operations are
+						// catastrophic events and further use should not happen
+						operations = null;
+					}
+				}
+				finally {
+					lock.unlock();
+				}
+			}
+		}
+
+		return connection;
+	}
+
```

Comment:
What concerns me is the schema creation or validation may require a lot of time to be performed and when delayed it may be triggered by session.beginTransaction() call.